### PR TITLE
Add graph runner support with torch compile on CPU

### PR DESCRIPTION
**Pull request by CaoE**
## Motivation

Inspired by https://github.com/mingfeima/sglang/pull/73. We add CPU graph runner with torch compile to reduce python overhead to speed up decoding on CPU.

Profiling with disabling torch compile:
image

Profiling with enabling torch compile:
image

From the profiling results above, we can see that torch compile can reduce python overhead by reducing the module call stack.

## Modifications

* Fix the mutation errors in kernel registration on CPU.
* Register fake functions for custom kernels to support torch.compile on CPU.
* Convert custom parameters to native parameters of torch to avoid graph break on CPU.
* Add graph runner support on CPU. Note: cpu graph needs to be enabled using `--enable-torch-compile`

## Performance:

Enable compile vs. Disable compile:
image

## Accuracy:

  |   | enable torch compile | disable torch compile
-- | -- | -- | --
model id | dtype | mmlu | mmlu
meituan--DeepSeek-R1-Channel-INT8 | int8 | 0.87 | 0.871
Qwen--Qwen3-14B-FP8 | fp8 | 0.788 | 0.788
meta-llama--Llama-3.2-3B-Instruct | bf16 | 0.607 | 0.607
microsoft--Phi-4-multimodal-instruct | bf16 | 0.689 | 0.688

## Checklist

- [x] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [x] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [x] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [x] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).

### CODE REVIEW

**Comment by mingfeima** [comment]
@CaoE have we collected traces for gpus ？you can ask validation team they have all type of gpus.

**Comment by CaoE** [comment]
> @CaoE have we collected traces for gpus ？you can ask validation team they have all type of gpus.

Yes and I found that CUDA has handled such custom parameters in process_weights_after_loading while cpu not. https://github.com/sgl-project/sglang/blob/main/python/sglang/srt/layers/quantization/w8a8_int8.py#L338. 

I will follow such process to avoid graph break instead of expanding `BasevLLMParameter`.

File: python/sglang/srt/layers/parameter.py [hunk 1]
```diff
@@ -1,10 +1,12 @@
 """Adapted from https://github.com/vllm-project/vllm/blob/v0.6.4.post1/vllm/model_executor/parameter.py"""
 
+import copy
 import logging
 from fractions import Fraction
 from typing import Callable, Optional, Union
 
 import torch
+import torch.utils._pytree as pytree
```
Comment (mingfeima) [inline python/sglang/srt/layers/parameter.py]: where is this import used?
and also the one above.

File: python/sglang/srt/layers/quantization/w8a8_int8.py [hunk 2]
```diff
@@ -333,6 +333,7 @@ def __init__(self, quantization_config: W8A8Int8Config):
         self.quantization_config = quantization_config
 
     def process_weights_after_loading(self, layer: torch.nn.Module) -> None:
+        layer.weight_scale = Parameter(layer.weight_scale.data, requires_grad=False)
```
Comment (mingfeima) [inline python/sglang/srt/layers/quantization/w8a8_int8.py]:
```suggestion
```

File: python/sglang/srt/layers/quantization/w8a8_int8.py [hunk 3]
```diff
@@ -341,7 +342,6 @@ def process_weights_after_loading(self, layer: torch.nn.Module) -> None:
             return
 
         layer.weight = Parameter(layer.weight.t(), requires_grad=False)
-        layer.weight_scale = Parameter(layer.weight_scale.data, requires_grad=False)
```
Comment (mingfeima) [inline python/sglang/srt/layers/quantization/w8a8_int8.py:345]: keep this line and do this?
```
if _is_cpu:
   ## handle weight:
else:
  layer.weight = layer.weight.t()

layer.weight = Parameter(layer.weight, requires_grad=False)
layer.weight_scale = Parameter(layer.weight_scale.data, requires_grad=False)
```

File: python/sglang/srt/layers/quantization/w8a8_int8.py [hunk 4]
```diff
@@ -475,12 +482,6 @@ def process_weights_after_loading(self, layer: torch.nn.Module) -> None:
 
         layer.w13_weight = Parameter(layer.w13_weight, requires_grad=False)
         layer.w2_weight = Parameter(layer.w2_weight, requires_grad=False)
-        layer.w13_weight_scale = Parameter(
-            layer.w13_weight_scale.data, requires_grad=False
-        )
-        layer.w2_weight_scale = Parameter(
-            layer.w2_weight_scale.data, requires_grad=False
-        )
```
Comment (mingfeima) [inline python/sglang/srt/layers/quantization/w8a8_int8.py]: same as above. more neat to put the lines for the same ideas at the same place.

File: python/sglang/srt/managers/scheduler.py [hunk 5]
```diff
@@ -382,13 +382,16 @@ def __init__(
             avail_mem = get_available_gpu_memory(
```
Comment (mingfeima) [inline python/sglang/srt/managers/scheduler.py]: what would `get_available_gpu_memory` return for cpu device?

File: python/sglang/srt/managers/scheduler_metrics_mixin.py [hunk 6]
```diff
@@ -184,8 +184,9 @@ def log_decode_stats(
             msg += f"pre-allocated usage: {self.disagg_decode_prealloc_queue.num_tokens_pre_allocated / self.max_total_num_tokens:.2f}, "
             msg += f"#retracted-req: {len(self.disagg_decode_prealloc_queue.retracted_queue)}, "
 
+        graph_msg = f"cpu torch-compile" if self.device == "cpu" else "cuda graph"
         msg += (
-            f"cuda graph: {can_run_cuda_graph}, "
+            f"{graph_msg}: {can_run_graph}, "
```
Comment (mingfeima) [inline python/sglang/srt/managers/scheduler_metrics_mixin.py]:
```suggestion
 f"{'cpu torch-compile' if self.device == 'cpu' else 'cuda graph'}: {can_run_graph}, "
```
or we just print `sefl.device` + graph: use `cpu graph` or `cuda graph` for the naming, skip torch compile.

**Review by mingfeima — APPROVED** [review]
Nice work! I suggest we unify the naming rules in sglang for graph mode.

for example, gpu uses `cuda graph`, and for cpu we simply use `cpu graph` instead of `torch.compile`. Easier to understand the concepts for users.

and our position is that we implement CPUGraphRunner with torch.compile, right?

File: python/sglang/srt/layers/quantization/w8a8_int8.py (hunk 3 above)
Comment (CaoE) [reply python/sglang/srt/layers/quantization/w8a8_int8.py:345]: I Moved `layer.weight_scale = Parameter(layer.weight_scale.data, requires_grad=False)` to the begin of `process_weights_after_loading` because CPU also need this line.

Comment (CaoE) [reply python/sglang/srt/layers/quantization/w8a8_int8.py:345]: Revised .

File: python/sglang/srt/layers/parameter.py (hunk 1 above)
Comment (CaoE) [reply python/sglang/srt/layers/parameter.py]: Thanks for pointing out this. Removed.

File: python/sglang/srt/managers/scheduler.py (hunk 5 above)
Comment (CaoE) [reply python/sglang/srt/managers/scheduler.py]: https://github.com/sgl-project/sglang/blob/3d6be1fbce6c36dc090cdc8628f0ab88a0003b5a/python/sglang/srt/utils.py#L413 This supports cpu device.

File: python/sglang/srt/layers/quantization/w8a8_int8.py (hunk 4 above)
Comment (CaoE) [reply python/sglang/srt/layers/quantization/w8a8_int8.py]: Revised.

File: python/sglang/srt/managers/scheduler_metrics_mixin.py (hunk 6 above)
Comment (CaoE) [reply python/sglang/srt/managers/scheduler_metrics_mixin.py]: Use `f"{'cpu graph' if self.device == 'cpu' else 'cuda graph'}: {can_run_graph}, "` instead.

**Comment by CaoE** [comment]
> and our position is that we implement CPUGraphRunner with torch.compile, right?

Yes. We implement CPUGraphRunner with torch.compile.

File: python/sglang/srt/managers/scheduler_metrics_mixin.py (hunk 6 above)
Comment (mingfeima) [reply python/sglang/srt/managers/scheduler_metrics_mixin.py]: f"{self.device} graph: {can_run_graph}, "

Comment (CaoE) [reply python/sglang/srt/managers/scheduler_metrics_mixin.py]: Use `cpu graph of torch.compile` to distinguish it from `cuda graph`.  Unlike cuda graph, cpu graph depends on torch compile, while torch compile is an option on cuda graph.

**Comment by mingfeima** [comment]
@Alcanderian could you please help review this one?

File: python/sglang/srt/managers/scheduler_metrics_mixin.py [hunk 7]
```diff
@@ -185,7 +185,7 @@ def log_decode_stats(
             msg += f"#retracted-req: {len(self.disagg_decode_prealloc_queue.retracted_queue)}, "
 
         msg += (
-            f"cuda graph: {can_run_cuda_graph}, "
+            f"{'cpu graph of torch.compile' if self.device == 'cpu' else 'cuda graph'}: {can_run_graph}, "
```
Comment (Alcanderian) [inline python/sglang/srt/managers/scheduler_metrics_mixin.py]: The message is too long for logging, can we just use `cpu graph`?

File: python/sglang/srt/model_executor/model_runner.py [hunk 8]
```diff
@@ -1590,31 +1599,38 @@ def init_double_sparsity_channel_config(self, selected_channel):
...
             return
 
         tic = time.perf_counter()
         before_mem = get_available_gpu_memory(self.device, self.gpu_id)
         logger.info(
-            f"Capture cuda graph begin. This can take up to several minutes. avail mem={before_mem:.2f} GB"
-        )
-        self.graph_runner = (
-            CudaGraphRunner(self) if not _is_npu else NPUGraphRunner(self)
+            f"Capture {'cpu graph of torch.compile' if self.device == 'cpu' else 'cuda graph'} begin. This can take up to several minutes. avail mem={before_mem:.2f} GB"
```
Comment (Alcanderian) [inline python/sglang/srt/model_executor/model_runner.py]: Just log `cpu graph`

File: python/sglang/srt/model_executor/model_runner.py [hunk 9]
```diff
@@ -1590,31 +1599,38 @@ def init_double_sparsity_channel_config(self, selected_channel):
...
         )
+        if device == "cuda":
```
Comment (Alcanderian) [inline python/sglang/srt/model_executor/model_runner.py]: Could we build a dict here?

File: python/sglang/srt/model_executor/model_runner.py [hunk 10]
```diff
@@ -1590,31 +1599,38 @@ def init_double_sparsity_channel_config(self, selected_channel):
                 .cuda()
             )
 
-    def init_device_graphs(self):
-        """Capture cuda graphs."""
+    def init_device_graphs(self, device="cuda"):
```
Comment (Alcanderian) [inline python/sglang/srt/model_executor/model_runner.py]: We can use self.device

File: python/sglang/srt/managers/scheduler_metrics_mixin.py (hunk 7 above)
Comment (mingfeima) [reply python/sglang/srt/managers/scheduler_metrics_mixin.py]: sure, np.

Comment (mingfeima) [reply python/sglang/srt/managers/scheduler_metrics_mixin.py]: @CaoE

Comment (CaoE) [reply python/sglang/srt/managers/scheduler_metrics_mixin.py]: Thanks for your comments. Used `cpu graph` instead.

File: python/sglang/srt/model_executor/model_runner.py (hunk 8 above)
Comment (CaoE) [reply python/sglang/srt/model_executor/model_runner.py]: Used `cpu` graph instead.

File: python/sglang/srt/model_executor/model_runner.py (hunk 9 above)
Comment (CaoE) [reply python/sglang/srt/model_executor/model_runner.py]: Used dict.

File: python/sglang/srt/model_executor/model_runner.py (hunk 10 above)
Comment (CaoE) [reply python/sglang/srt/model_executor/model_runner.py]: Used ` self.device`.

File: python/sglang/srt/model_executor/model_runner.py [hunk 11]
```diff
@@ -1591,30 +1600,42 @@ def init_double_sparsity_channel_config(self, selected_channel):
...
+        if self.device == "cpu" and not self.server_args.enable_torch_compile:
             return
 
         tic = time.perf_counter()
         before_mem = get_available_gpu_memory(self.device, self.gpu_id)
         logger.info(
-            f"Capture cuda graph begin. This can take up to several minutes. avail mem={before_mem:.2f} GB"
-        )
-        self.graph_runner = (
-            CudaGraphRunner(self) if not _is_npu else NPUGraphRunner(self)
```
Comment (CaoE) [inline python/sglang/srt/model_executor/model_runner.py:1623]: Hi @Alcanderian I want to confirm if self.device is "cuda" is it possible that `self.graph_runner = NPUGraphRunner(self)`? Is NPUGraphRunner also a kind of `CudaGraphRunner` so we also call it "cuda graph" in the log message?

```
graph_runners = {
    "cuda": CudaGraphRunner, # or we should keep CudaGraphRunner if not _is_npu else NPUGraphRunner,
    "npu": CudaGraphRunner if not _is_npu else NPUGraphRunner,
    "cpu": CPUGraphRunner,
}
```

Comment (Alcanderian) [reply python/sglang/srt/model_executor/model_runner.py:1623]: AMD also use cuda, so we should use a default_dict whose default value is `CUDAGraphRunner`, and only define 2 elements
```
"npu": NPUGraphRunner,
"cpu": CPUGraphRunner, 
```

Comment (CaoE) [reply python/sglang/srt/model_executor/model_runner.py:1623]: Fixed. Thanks~

**Comment by mingfeima** [comment]
@CaoE need to solve the conflicts!

**Comment by CaoE** [comment]
> @CaoE need to solve the conflicts!

Fixed the conflicts.

**Comment by FlamingoPg** [comment]
CI triggered, thx

File: python/sglang/srt/model_executor/model_runner.py [hunk 12]
```diff
@@ -1593,30 +1603,39 @@ def init_double_sparsity_channel_config(self, selected_channel):
...
-            f"Capture cuda graph begin. This can take up to several minutes. avail mem={before_mem:.2f} GB"
+            f"Capture {'cpu graph' if self.device == 'cpu' else 'cuda graph'} begin. This can take up to several minutes. avail mem={before_mem:.2f} GB"
         )
-        self.graph_runner = (
-            CudaGraphRunner(self) if not _is_npu else NPUGraphRunner(self)
+        graph_runners = defaultdict(
+            lambda: CudaGraphRunner,
+            {
+                "cpu": CPUGraphRunner,
+                "npu": CudaGraphRunner if not _is_npu else NPUGraphRunner,
```
Comment (Alcanderian) [inline python/sglang/srt/model_executor/model_runner.py]:
```suggestion
                "npu": NPUGraphRunner,
```

**Review by Alcanderian — APPROVED** [review]

**Comment by CaoE** [comment]
@mingfeima Could you please help trigger CI ?

**Comment by CaoE** [comment]
Hi @Alcanderian @zhyncs Can this PR be landed since the NPU graph runner has been re-landed ? Recently, some AMD-GPU and NPU-related tests have failed, most of which seem unrelated to this PR. Thanks.

**Comment by Alcanderian** [comment]
> Hi @Alcanderian @zhyncs Can this PR be landed since the NPU graph runner has been re-landed ? Recently, some AMD-GPU and NPU-related tests have failed, most of which seem unrelated to this PR. Thanks.

please fix xeon CI

**Comment by CaoE** [comment]
> please fix xeon CI

Fixed https://github.com/sgl-project/sglang/actions/runs/17227399712/job/48876607152?pr=7843.

**Review by FlamingoPg — APPROVED** [review]

**Comment by CaoE** [comment]
Hi @FlamingoPg @Alcanderian Can this PR be merged?

**Comment by CaoE** [comment]
@FlamingoPg @Alcanderian https://github.com/sgl-project/sglang/actions/runs/17231906249/job/48964007907?pr=7843 doesn't seem to be related to this PR.

**Comment by CaoE** [comment]
Hi @zhyncs Could you please help merge this PR ? Thank you.

**Comment by FlamingoPg** [comment]
> Hi @zhyncs Could you please help merge this PR ? Thank you.

@zhyncs LGTM, any other comment?

File: python/sglang/srt/managers/scheduler.py [hunk 13]
```diff
@@ -181,7 +181,7 @@ class GenerationBatchResult:
     extend_input_len_per_req: List[int]
     extend_logprob_start_len_per_req: List[int]
     bid: int
-    can_run_cuda_graph: bool
```
Comment (zhyncs) [inline python/sglang/srt/managers/scheduler.py]: can we keep the old name

Comment (CaoE) [reply python/sglang/srt/managers/scheduler.py]: Changing `can_run_cuda_graph` to `can_run_graph` is simply to indicate that graph is also supported on the CPU, thus reducing confusion. Depending on the device, it can be determined whether it is a CUDA graph or a CPU graph. Do you think we should keep all `can_run_cuda_graph`? Are there any unconsidered implications of changing can_run_cuda_graph to can_run_graph? This is fine for us, and I can change all `can_run_graph` back to `can_run_cuda_graph`.

Comment (zhyncs) [reply python/sglang/srt/managers/scheduler.py]: I hope to minimize the impact on downstream forks as much as possible, usually new hardware changes are best made independently, with minimal changes to existing NVIDIA GPUs.

Comment (CaoE) [reply python/sglang/srt/managers/scheduler.py]: Thanks for your comments. I'll change them back.

Comment (CaoE) [reply python/sglang/srt/managers/scheduler.py]: Modified.

**Comment by CaoE** [comment]
Hi @zhyncs https://github.com/sgl-project/sglang/actions/runs/17372540433/job/49363628357?pr=7843 doesn't seem to be related to this PR. Can this PR be merged?

**Comment by CaoE** [comment]
Hi @zhyncs Could you please take another look? Thank you.
